lib/core/src/Cardano/Wallet/Primitive/BlockSummary.hs

```diff
@@ -123,17 +132,59 @@ toAscBlockEvents (ChainEvents bs) = Map.elems bs
 data BlockEvents = BlockEvents
     { slot :: !Slot
     , blockHeight :: !(Quantity "block" Word32)
-    , transactions :: [(Int, Tx)]
+    , transactions :: Sublist Tx
         -- ^ (Index of the transaction within the block, transaction data)
         -- INVARIANT: The list is ordered by ascending index.
-    , delegations :: [(Int, DelegationCertificate)]
+    , delegations :: Sublist DelegationCertificate
         -- ^ (Index of the delegation within the block, delegation data)
         -- INVARIANT: The list is ordered by ascending index.
     } deriving (Eq, Ord, Generic, Show)
 
+-- | A data type representing a sublist of a total list.
+-- Such a sublist typically arises by filtering and keeps
+-- track of the indices of the filtered list elements.
+--
+-- The main purpose of this data type is optimization:
+-- When processing whole 'Block', we want to avoid copying
+-- and redecorating the entire list of transactions in that 'Block';
+-- instead, we want to copy a pointer to this list.
+data Sublist a = All [a] | Some [(Int, a)]
+    deriving (Eq, Ord, Show)
+
+-- | Construct a 'Sublist' representing the whole list.
+wholeList :: [a] -> Sublist a
+wholeList = All
+
+-- | Construct a 'Sublist' from a list of indexed items.
+unsafeMkSublist :: [(Int,a)] -> Sublist a
+unsafeMkSublist = Some
+
+-- | Filter a 'Sublist' by a predicate.
+filterSublist :: (a -> Bool) -> Sublist a -> Sublist a
+filterSublist p (All xs) = filterSublist p $ Some $ zip [0..] xs
+filterSublist p (Some ixs) = Some [ ix | ix <- ixs, p (snd ix) ]
+
+instance Functor Sublist where
+    fmap f (All xs) = All (map f xs)
+    fmap f (Some ixs) = Some [ (i, f x) | (i,x) <- ixs ]
+
+instance Foldable Sublist where
+    foldr f b = foldr f b . toList
+    null = null . toList
+    toList (All as) = as
+    toList (Some ixs) = map snd ixs
+
+-- | Returns 'True' if the 'BlockEvents' contains empty
+-- 'transactions' and 'delegations'.
 nullBlockEvents :: BlockEvents -> Bool
-nullBlockEvents BlockEvents{transactions=[],delegations=[]} = True
-nullBlockEvents _ = False
+nullBlockEvents BlockEvents{transactions,delegations}
+    = null transactions && null delegations
+
+-- | Merge two 'Sublist' assuming that they are sublists of the /same/ list.
```

Comment:
yeah, when sublists of the same list are considered then this is perfectly true. The question is can we outlaw calling this function with sublists of different lists?  Maybe it would be worth adding some invariant avalilable for export here checking this? Possible?